--- a/setup.cfg
+++ b/setup.cfg
@@ -107,7 +107,7 @@ extension =
     S=tools.flake8_plugin:SentryCheck
 
 [bdist_wheel]
-python-tag = py38
+python-tag = py311
 
 [coverage:run]
 omit =
--- a/src/sentry/utils/distutils/commands/base.py
+++ b/src/sentry/utils/distutils/commands/base.py
@@ -122,7 +122,7 @@ class BaseBuildCommand(Command):
 
         if node_version[2] is not None:
             log.info(f"using node ({node_version})")
-            self._run_command(["yarn", "install", "--production", "--frozen-lockfile", "--quiet"])
+            self._run_command(["yarn", "install", "--production", "--frozen-lockfile"])
 
     def _run_command(self, cmd, env=None):
         cmd_str = " ".join(cmd)
