--- a/snuba/clickhouse/http.py
+++ b/snuba/clickhouse/http.py
@@ -292,7 +292,7 @@ class HTTPBatchWriter(BatchWriter[bytes]):
         chunk_size: Optional[int] = None,
         buffer_size: int = 0,
     ):
-        self.__pool = HTTPConnectionPool(host, port)
+        self.__pool = HTTPConnectionPool(host, port, maxsize=2)
         self.__executor = ThreadPoolExecutor()
         self.__metrics = metrics
 
--- a/snuba/settings/__init__.py
+++ b/snuba/settings/__init__.py
@@ -425,6 +425,7 @@ def _load_settings(obj: MutableMapping[str, Any] = locals()) -> None:
     provide a full absolute path such as `/foo/bar/my_settings.py`."""
 
     import importlib
+    import importlib.abc
     import importlib.util
     import os
 
--- a/snuba/utils/uwsgi.py
+++ b/snuba/utils/uwsgi.py
@@ -4,20 +4,6 @@ import os
 import sys
 from typing import MutableMapping, NoReturn
 
-PYUWSGI_PROG = """\
-import os
-import sys
-
-orig = sys.getdlopenflags()
-sys.setdlopenflags(orig | os.RTLD_GLOBAL)
-try:
-    import pyuwsgi
-finally:
-    sys.setdlopenflags(orig)
-
-pyuwsgi.run()
-"""
-
 
 def _prepare_environ(
     options: dict[str, bool | int | str | None], env: MutableMapping[str, str]
@@ -71,6 +57,4 @@ def run(module: str, bind: str, **kwargs: bool | int | str | None) -> NoReturn:
 
     _prepare_environ(options, os.environ)
 
-    # TODO: https://github.com/lincolnloop/pyuwsgi-wheels/pull/17
-    cmd = (sys.executable, "-c", PYUWSGI_PROG)
-    os.execvp(cmd[0], cmd)
+    os.execvp("uwsgi", ("uwsgi",))
