--- a/snuba/cli/admin.py
+++ b/snuba/cli/admin.py
@@ -35,9 +35,11 @@ def admin(
         WSGIRequestHandler.protocol_version = "HTTP/1.1"
         application.run(host=host, port=port, threaded=True, debug=debug)
     else:
+        import mywsgi
+
         if log_level:
             os.environ["LOG_LEVEL"] = log_level
-        uwsgi.run(
+        mywsgi.run(
             "snuba.admin.wsgi:application",
             f"{host}:{port}",
             processes=processes,
--- a/snuba/cli/api.py
+++ b/snuba/cli/api.py
@@ -5,7 +5,6 @@ import click
 
 from snuba.attribution.log import flush_attribution_producer
 from snuba.environment import setup_logging
-from snuba.utils import uwsgi
 
 
 @click.command()
@@ -47,11 +46,13 @@ def api(
         WSGIRequestHandler.protocol_version = "HTTP/1.1"
         application.run(host=host, port=port, threaded=True, debug=debug)
     else:
+        import mywsgi
+
         if log_level:
             os.environ["LOG_LEVEL"] = log_level
 
         with flush_attribution_producer():
-            uwsgi.run(
+            mywsgi.run(
                 "snuba.web.wsgi:application",
                 f"{host}:{port}",
                 processes=processes,
--- a/snuba/settings/__init__.py
+++ b/snuba/settings/__init__.py
@@ -349,6 +349,7 @@ def _load_settings(obj: MutableMapping[str, Any] = locals()) -> None:
     provide a full absolute path such as `/foo/bar/my_settings.py`."""
 
     import importlib
+    import importlib.abc
     import importlib.util
     import os
 
